packages/@react-stately/numberfield/src/useNumberFieldState.ts

```diff
@@ -11,133 +11,298 @@
+  props: NumberFieldProps
 ): NumberFieldState {
-  let {minValue, maxValue, step = 1, formatOptions, value, defaultValue, onChange} = props;
+  let {locale} = useLocale();
+  let {minValue = Number.MIN_SAFE_INTEGER, maxValue = Number.MAX_SAFE_INTEGER, step, formatOptions, value, defaultValue, onChange} = props;
+  let [currentNumeralSystem, setCurrentNumeralSystem] = useState<string | undefined>();
+
+  let numeralOverride;
+  switch (currentNumeralSystem) {
+    case 'arab':
+      numeralOverride = 'u-nu-arab';
+      break;
+    case 'hanidec':
+      numeralOverride = 'u-nu-hanidec';
+      break;
+    case 'latin':
+      numeralOverride = 'u-nu-latn';
+      break;
+    default:
+      numeralOverride = '';
+      break;
+  }
+  let inputValueFormatter = useNumberFormatter(formatOptions, numeralOverride);
+  let numberParser = useNumberParser(inputValueFormatter, numeralOverride);
+  let intlOptions = useMemo(() => inputValueFormatter.resolvedOptions(), [inputValueFormatter]);
+
+  let isMaxRange = minValue === Number.MIN_SAFE_INTEGER && maxValue === Number.MAX_SAFE_INTEGER;
+
+  let {minusSign, plusSign, decimal, validCharacters, literals, group} = numberParser.symbols;
+  minusSign = minValue >= 0 || !minusSign ? '' : minusSign;
+  plusSign = maxValue <= 0 || !plusSign ? '' : plusSign;
+
+  // javascript doesn't recognize NaN === NaN, so multiple onChanges will get fired if we don't ignore consecutive ones
+  // in addition, if the input starts with a number, then we'll count that as the last val dispatched, we only need to calculate it the first time
+  let startingValue = useMemo(() => {
+    if (!isNaN(value)) {
+      return value;
+    } else if (!isNaN(defaultValue)) {
+      return defaultValue;
+    }
+    return NaN;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+  let lastValDispatched = useRef(startingValue);
+  let smartOnChange = useCallback((val) => {
+    if (!isNaN(val) || !isNaN(lastValDispatched.current)) {
+      onChange?.(val);
+    }
+    lastValDispatched.current = val;
+  }, [lastValDispatched, onChange]);
 
-  const numberParser = useNumberParser();
-  const textValueFormatter = useNumberFormatter(formatOptions);
-  const inputValueFormatter = useNumberFormatter();
 
-  const [numberValue, setNumberValue] = useControlledState<number>(value, defaultValue || 0, onChange);
-  let initialInputValue = inputValueFormatter.format(numberValue);
-  const [inputValue, setInputValue] = useState(isNaN(value) && isNaN(defaultValue) ? '' : initialInputValue);
-  const [isValid, setIsValid] = useState(isInputValueValid(numberValue, maxValue, minValue));
+  let [numberValue, setNumberValue] = useControlledState<number>(value, isNaN(defaultValue) ? NaN : defaultValue, smartOnChange);
 
-  const minusSign = useRef('-');
+  let initialInputValue = isNaN(numberValue) ? '' : inputValueFormatter.format(numberValue);
+  let [inputValue, setInputValue] = useState(initialInputValue);
```

Comment:
You could use the function form for `useState` to avoid recomputing the initial input value on each render. Though I guess you already compute it below anyway. Maybe you could reuse `textValue`?